### PR TITLE
[gpt_oss_d_p] GPT-OSS prefill: MoE wrapper (router + fused experts, SwiGLU-OAI)

### PR DESCRIPTION
**Pull request by mdragulaTT**
> **P3 of the GPT-OSS prefill stack.** #50223 (attention) + #50228 (KV cache) are merged, so this now stacks directly on `main`; the diff is the MoE delta only.

## What
The **MoE wrapper** — a thin `TtGptOssMoE` composing the DeepSeek EP substrate, plus a gpt-oss-specific router. Mirrors `minimax_m3`'s `TtMiniMaxMoE`. Every gpt-oss layer is MoE (no shared expert, no dense branch).

## This PR
- **`tt/moe/router.py`** — `TtGptOssRouter`, ported from `gpt_oss/tt/topk.py`: `linear(+bias) → topk(4) → softmax over the 4 selected`. gpt-oss selects top-4 *then* softmaxes (DeepSeek's gate scores all experts first), so routing is external — not DeepSeek's `TtMoEGatePrefill`.
- **`tt/moe/tt_gpt_oss_moe.py`** — `TtGptOssMoE`, mirrors `TtMiniMaxMoE` **minus the internal gate and the shared expert**; reuses `deepseek_v3_d_p.tt.moe.*` dispatch/combine/reduce/routing_setup + `TtRoutedExpert` (`activation=SwiGluOai`, `combine init_zeros=True`) verbatim.
- **`tt/moe/weights.py`** — expert weight prep: de-interleave `gate_up_proj[...,::2]`/`[...,1::2]` (+ matching bias split), transpose to HF `(out,in)`, global expert-id order.
- **`tt/mlp.py`** — thin router + MoE wrapper (all-MoE, no shared expert, no dense branch).

## Expert bias (#49619, merged)
gpt-oss experts carry gate/up/down **biases**. #49619 added bias support to `TtRoutedExpert` / the fused `unified_routed_expert_moe` kernel (gate/up bias before the clamp, down bias after the down matmul; SwiGluOai only), and it is **merged**. This PR runs **biased** (`use_expert_bias=True`): `weights.py` preps the per-expert biases and `TtGptOssMoE` passes them straight through to `TtRoutedExpert`. `use_expert_bias=False` keeps the bias-free path available.

## Tests (`tests/unit/test_moe_vs_ref.py`)
- **`test_router_vs_ref`** (single card): topk indices vs ref — mismatches proven to be genuine bf16 top-4/5 near-ties; softmax weights PCC **0.9955**.
- **`test_routed_expert_ffn_vs_ref`** (single card) — parametrized **`[biasfree, biased]`**: SwiGLU-OAI expert FFN vs a torch reference, both bias-free and with per-expert gate/up/down biases (the deployment default). min PCC **0.9833** (both variants). The `biased` case exercises the exact path the model runs.
- **`test_gpt_oss_moe_vs_ref`** (galaxy-gated): full router→dispatch→combine→reduce at EP=32 — auto-skips on a single card.

## Follow-ups
- Add the full-MoE + per-layer golden-KV validation on a 4×8 Blackhole galaxy (P5).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mdragula/gpt-oss-moe)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mdragula/gpt-oss-moe)

### CODE REVIEW

File: models/demos/gpt_oss_d_p/tt/moe/tt_gpt_oss_moe.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            "TtGptOssMoE requires external topk_indices/topk_weights (from TtGptOssRouter); " "it has no internal gate."
+        )
+        indices, scores = topk_indices, topk_weights
+
+        tt_expert_offsets, tt_expert_token_counts, tt_expert_region_offsets, _ = self.routing_setup(
+            ttnn_top_k_experts_indices=indices,
+            num_routed_experts=self.num_routed_experts,
+            seq_len_per_chip=self.seq_len_per_chip,
+            num_experts_per_tok=self.num_experts_per_tok,
+        )
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/moe/tt_gpt_oss_moe.py:181]: Done (b82a68c). Removed the `seq_len_per_chip=` kwarg. `TtMoERoutingSetup.forward` only accepts `(ttnn_top_k_experts_indices, num_routed_experts, num_experts_per_tok)`; the extra kwarg slipped in when the deepseek signature changed under a rebase, and the minimax_m3 template does not pass it.

File: models/demos/gpt_oss_d_p/tests/unit/test_moe_vs_ref.py [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+    This is the ONLY remaining wiring gap for the full EP MoE; router + routed-expert plumbing are
+    already validated by the two single-card tests above."""
+    from models.demos.gpt_oss_d_p.tt.config import MeshConfig
+    from models.demos.gpt_oss_d_p.tt.mlp import MLP
+    from models.demos.gpt_oss_d_p.utils.general_utils import get_default_num_links
+
+    try:
+        from models.demos.gpt_oss_d_p.tt.ccl import CCLManager  # not yet implemented in this package
+    except ImportError:
+        pytest.skip("gpt_oss_d_p CCL manager not yet implemented; full EP MoE test is blocked on it.")
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tests/unit/test_moe_vs_ref.py]: The full `TtGptOssMoE` forward is fabric-based EP dispatch/combine, so it needs the multi-device fabric: a single-card build hits `un-initialized fabric context`, so it cannot run on 1 chip. It is galaxy-validated via the runtime harness, and `CCLManager` lands in the runtime PR (#50265), so the import + skip here is the intended stacking behavior (docstring refreshed). To cover the host-side layout math the galaxy test relies on, added a pure-torch `test_prepare_routed_expert_weights_deinterleave`.

File: models/demos/gpt_oss_d_p/tt/mlp.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            ep_seq_len_per_chip, E, hf_config.num_experts_per_tok, mesh_device.get_num_devices(), dgs, 2
+        )
+
+        # De-interleave gate_up_proj, transpose to HF (out,in), global expert order 0..E-1.
+        # Biases are kept SEPARATE (not attached to the weight dicts) and passed through to
+        # TtRoutedExpert (#49619 merged). None in cache-only mode -> TtRoutedExpert loads tilized
+        # weights from cache.
+        cache_only = not state_dict
+        if cache_only:
+            routed_w, routed_b = None, None
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/mlp.py]: Good catch, fixed (b82a68c). `TtRoutedExpert` does not persist biases to the TTNN weight cache, so cache-only builds now reload them from a small sidecar written on the real-weights build, and fail loud if it is missing. No more silent bias-free run.

File: models/demos/gpt_oss_d_p/tests/unit/test_moe_vs_ref.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+        return out
+
+    x = torch.randn(rows, seq_len, H, dtype=torch.bfloat16)
+    ref = [_ref_moe(x[r].float()) for r in range(rows)]
+
+    # Build the HF GptOssMLP state dict: router.{weight,bias} + interleaved experts.* .
+    gate_up = torch.zeros(E, H, 2 * I)
+    gate_up[..., ::2] = torch.stack([w["gate_proj"].t() for w in experts])  # (in=H, out=I)
+    gate_up[..., 1::2] = torch.stack([w["up_proj"].t() for w in experts])
+    gate_up_bias = torch.zeros(E, 2 * I)  # bias-free ref -> zeros
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tests/unit/test_moe_vs_ref.py:476]: Addressed (b82a68c): `test_routed_expert_ffn_vs_ref` now runs a `[biased]` variant on-card (min PCC 0.983), and the new pure-torch `test_prepare_routed_expert_weights_deinterleave` exercises the non-zero gate/up/down bias split through `prepare_routed_expert_weights`. A full biased MLP-through-prep check at galaxy scale is future work.

File: models/demos/gpt_oss_d_p/tt/moe/weights.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        num_experts: number of routed experts E (global).
+        hidden_size: model embedding dim (hidden).
+        intermediate_size: MoE FFN hidden dim (inter).
+
+    Returns:
+        (routed_expert_weights, routed_expert_biases):
+          * routed_expert_weights: list[dict] len E, global order 0..E-1, keys
+            ``gate_proj``/``up_proj``/``down_proj`` in HF ``(out, in)`` layout.
+          * routed_expert_biases: list[dict] len E, global order 0..E-1, keys
+            ``gate_bias``/``up_bias``/``down_bias`` (1D). SEPARATE from the weights (#49619).
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/moe/weights.py]: Done (b82a68c). Docstring now lists `gate_proj_bias`/`up_proj_bias`/`down_proj_bias`, the actual returned keys.

File: models/demos/gpt_oss_d_p/tt/moe/tt_gpt_oss_moe.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        num_routed_experts: int,
+        num_experts_per_tok: int,
+        metadata_len: int,
+        max_dispatched_tokens_per_expert: int,
+        max_dispatch_buffer_token_size: int,
+        seq_len_per_chip: int,
+        emb_dim: int,
+        hidden_dim: int,
+        routed_expert_weights: list,  # per-expert {gate_proj, up_proj, down_proj}, global id order 0..E-1
+        routed_expert_biases: list = None,  # per-expert {gate_bias, up_bias, down_bias}, same order (#49619)
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply models/demos/gpt_oss_d_p/tt/moe/tt_gpt_oss_moe.py]: Done (b82a68c). The param comment now lists `gate_proj_bias`/`up_proj_bias`/`down_proj_bias`.

**Review by jmalone-tt — APPROVED** [review]
